src/replication.c

```diff
@@ -125,37 +127,330 @@ static int handle_mgmt_event_fd(replica_t *replica);
+	BUILD_REPLICA_MGMT_HDR(rmgmtio, mgmt_opcode, data_len);
+
+	mgmt_cmd->io_hdr = rmgmtio;
+	mgmt_cmd->io_bytes = 0;
+	mgmt_cmd->data = buf;
+	mgmt_cmd->mgmt_cmd_state = WRITE_IO_SEND_HDR;
+
+	MTX_LOCK(&replica->r_mtx);
+	TAILQ_INSERT_TAIL(&replica->mgmt_cmd_queue, mgmt_cmd, mgmt_cmd_next);
+	MTX_UNLOCK(&replica->r_mtx);
+
+	if (write(replica->mgmt_eventfd1, &num, sizeof (num)) !=
+	    sizeof (num)) {
+		REPLICA_NOTICELOG("Failed to inform to mgmt_eventfd "
+		    "for replica(%p)\n", replica);
+		return (ret = -1);
+	}
+	REPLICA_LOG("start_rebuild opcode sent for Replica ip:%s port:%d "
+	    "state:%d\n", replica->ip, replica->port, replica->state);
+	return ret;
+}
+
+/* Rebuild can be of two type:-
+ * 1. Rebuilding a replica from healthy replica
+ * 2. Rebuilding a replica from all other replica(Mesh rebuild)
+ * API does the following tasks:-
+ * 1. find if there is healthy replica available for rebuilding
+ * 2. find downgraded replica for rebuild
+ * 3. Call prepare_for_rebuild to on healthy replica or on all
+ * other replicas to get IP address and port for replicas.
+ */
+void
+trigger_rebuild(spec_t *spec)
+{
+	int ret = 0;
+	uint64_t max = 0;
+	struct timespec now, diff;
+	replica_t *replica = NULL;
+	replica_t *target_replica = NULL;
+	replica_t *healthy_replica = NULL;
+
+	if (spec->rebuild_in_progress == true) {
+		REPLICA_NOTICELOG("Rebuild is already in progress "
+		    "on volume:%s\n", spec->volname);
+		MTX_UNLOCK(&spec->rq_mtx);
+		return;
+	}
+
+	if (spec->ready != true) {
+		REPLICA_NOTICELOG("Volume:%s is not ready to accept IOs\n",
+		    spec->volname);
+		MTX_UNLOCK(&spec->rq_mtx);
+		return;
+	}
+
+	if (!spec->degraded_rcount) {
+		REPLICA_NOTICELOG("No downgraded replica on volume:%s "
+		", rebuild will not be attempted\n", spec->volname);
+		return;
+	}
+
+	clock_gettime(CLOCK_MONOTONIC, &now);
+
+	TAILQ_FOREACH(replica, &spec->rq, r_next) {
+		/* Find healthy replica */
+		if (replica->state == ZVOL_STATUS_HEALTHY) {
+			if (healthy_replica == NULL)
+				healthy_replica = replica;
+			continue;
+		}
+
+		timesdiff(replica->create_time, now, diff);
+		if (diff.tv_sec <= replica_timeout) {
+			REPLICA_LOG("Replica:%p added very recently, "
+			    "skipping rebuild on it\n", replica);
+			continue;
+		}
 
-	if(((spec->healthy_rcount + spec->degraded_rcount >= spec->consistency_factor) &&
-		(spec->healthy_rcount >= 1))||
-		(spec->healthy_rcount  + spec->degraded_rcount 
-			>= MAX(spec->replication_factor - spec->consistency_factor + 1, spec->consistency_factor))) {
-		if (spec->ready == false)
-		{
-			max = 0;
-			TAILQ_FOREACH(replica, &spec->rq, r_next)
-				max = (max < replica->initial_checkpointed_io_seq) ?
-				    replica->initial_checkpointed_io_seq : max;
+		if (max < replica->initial_checkpointed_io_seq) {
+			max = replica->initial_checkpointed_io_seq;
+			target_replica = replica;
+		}
 
+		/*
+		 * Increase io_seq only when we do not have healthy replica
+		 */	
+		if (healthy_replica == NULL) {
```

Comment:
we don't need this here now..

Reply:
basically, updating io_seq of spec..

Reply:
Done sire! Check if target_replica is null then return from here.